packages/headless-react/src/search-engine.tsx

```diff
@@ -0,0 +1,108 @@
+import {Controller, CoreEngine, SearchEngine} from '@coveo/headless';
+import {
+  ControllerDefinitionsMap,
+  InferControllerFromDefinition,
+  SearchEngineDefinitionOptions,
+  defineSearchEngine as defineBaseSearchEngine,
+  mapObject,
+} from '@coveo/headless/ssr';
+import {
+  Context,
+  createContext,
+  useContext,
+  useCallback,
+  useMemo,
+  useSyncExternalStore,
+} from 'react';
+import {
+  ContextCSRState,
+  ContextState,
+  ControllerHook,
+  InferControllerHooksMapFromDefinition,
+  ReactSearchEngineDefinition,
+} from './types';
+
+export function capitalize(s: string) {
+  return `${s.slice(0, 1).toUpperCase()}${s.slice(1)}`;
+}
+
+function isCSRContext<
+  TEngine extends CoreEngine,
+  TControllers extends ControllerDefinitionsMap<TEngine, Controller>,
+>(
+  ctx: Exclude<ContextState<TEngine, TControllers>, null>
+): ctx is ContextCSRState<TEngine, TControllers> {
+  return 'engine' in ctx;
+}
+
+function buildControllerHook<
+  TEngine extends CoreEngine,
+  TControllers extends ControllerDefinitionsMap<TEngine, Controller>,
+  TKey extends keyof TControllers,
+>(
+  context: Context<ContextState<TEngine, TControllers>>,
+  key: TKey
+): ControllerHook<InferControllerFromDefinition<TControllers[TKey]>> {
+  return () => {
+    const ctx = useContext(context);
+    if (ctx === null) {
+      throw 'Missing SSR state or CSR provider.';
```

Comment:
Since those will appear for implementers, I think we should limit the amount of acronyms in use.

Also, if that happens what (if anything) could an implementer do to potentially troubleshoot and fix the problem ? 

If there's something that can be done, the message should try to explain it.